README.md
# jquery-simple-tree

A jquery plugin for simple tree ui.

## Dependencies

* jquery

## Installation

Install from npm:

    $ npm install @kanety/jquery-simple-tree --save

## Usage

Build tree using list elements as follows:

```html
<ul id="tree">
  <li data-node-id="1">
    <span>text of 1</span>
    <ul>
      <li data-node-id="1.1">
        <span>text of 1.1</span>
        <ul>
          <li data-node-id="1.1.1">
            <span>text of 1.1.1</span>
          </li>
          <li data-node-id="1.1.2">
            <span>text of 1.1.2</span>
          </li>
        </ul>
      </li>
    </ul>
  </li>
</ul>
```

Then run:

```javascript
$('#tree').simpleTree();
```

### Options

Specify opened nodes:

```javascript
$('#tree').simpleTree({
  opened: [1, 1.1]
});
```

Save node state to `sessionStorage` or `localStorage`:

```javascript
$('#tree').simpleTree({
  store: 'session', // or 'local'
  storeKey: 'YOUR_KEY'
});
```

### Callbacks

Run callbacks when a node is opened or closed:

```javascript
$('#tree').simpleTree({
  ...
}).on('node:open', function(e, $node) {
  ...
}).on('node:close', function(e, $node) {
  ...
});
```

### Functions

Manipulate nodes from your script:

```javascript
var tree = $('#tree').data('simple-tree');  // get SimpleTree instance
tree.openByID(1);
tree.closeByID(1);
```

## License

The library is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG

## 0.5.0

* Add `node-keep-open` flag to keep node opened.
* Change css from `simple-tree-empty` to `simple-tree-hidden-icon`.
* Change data attributes from `node-lazy` to `node-has-children`.

## 0.4.0

* Support focus movement by keyboard.
* Change default icon template from span tag to a tag.

## 0.3.0

* Fix css class name.
* Bundle css with js.

## 0.2.1

* Fix jquery deprecated selector.

## 0.2.0

* Separate store class and change store options.
* Destroy existing instance on initialization.
* Change event hanlder class.
* Tweak css.

## 0.1.2

* Insert icon text before .tree-icon.

## 0.1.1

* Treat li without data-node-id as node.
* Tweak css style.

## 0.1.0

* First release.


test/jquery-sipmle-tree-spec.js
const NAMESPACE = 'simple-tree';

function keydown(elem, code) {
  $(elem).trigger($.Event('keydown', { keyCode: code }));
}

describe('jquery-simple-tree', () => {
  beforeEach(() => {
    document.body.innerHTML = __html__['index.html'];
    eval($('script').text());
  });

  describe('basic', () => {
    let $tree, $node, $icon;
    let $expander, $collapser;

    beforeEach(() => {
      $tree = $('#basic');
      $node = $tree.find('li[data-node-id="1"]');
      $icon = $node.children(`.${NAMESPACE}-icon`);
      $expander = $('#expander');
      $collapser = $('#collapser');
      $open1 = $('#open1');
      $close1 = $('#close1');
    });

    it('opens and closes', () => {
      $icon.click();
      expect($node.children('ul').is(':visible')).toEqual(false);
      $icon.click();
      expect($node.children('ul').is(':visible')).toEqual(true);
    });

    it('has expander and collapser', () => {
      $collapser.click();
      expect($tree.find('ul:visible').length).toEqual(1);
      $expander.click();
      expect($tree.find('ul:hidden').length).toEqual(0);
    });

    it('opens and closes by id', () => {
      $close1.click();
      expect($node.children('ul').is(':visible')).toEqual(false);
      $open1.click();
      expect($node.children('ul').is(':visible')).toEqual(true);
    });
  });

  describe('keyboard events', () => {
    let $tree, $node, $icon;

    beforeEach(() => {
      $tree = $('#basic');
      $node = $tree.find('li[data-node-id="1.1"]');
      $icon = $node.children(`.${NAMESPACE}-icon`);
    });

    it('moves focus by up key', () => {
      $tree.data(NAMESPACE).expand();
      keydown($icon, 38);
      expect($(document.activeElement).parent().attr('data-node-id')).toEqual('1');
    });

    it('moves focus by down key', () => {
      $tree.data(NAMESPACE).expand();
      keydown($icon, 40);
      expect($(document.activeElement).parent().attr('data-node-id')).toEqual('1.1.1');
    });

    it('moves focus by left key', () => {
      $tree.data(NAMESPACE).close($node);
      keydown($icon, 37);
      expect($(document.activeElement).parent().attr('data-node-id')).toEqual('1');
    });

    it('closes node by left key', () => {
      $tree.data(NAMESPACE).open($node);
      keydown($icon, 37);
      expect($node.find('ul').is(':visible')).toEqual(false);
    });

    it('moves focus by right key', () => {
      $tree.data(NAMESPACE).open($node);
      keydown($icon, 39);
      expect($(document.activeElement).parent().attr('data-node-id')).toEqual('1.1.1');
    });

    it('opens node by right key', () => {
      $tree.data(NAMESPACE).close($node);
      keydown($icon, 39);
      expect($node.find('ul').is(':visible')).toEqual(true);
    });
  });

  describe('open', () => {
    let $tree;

    beforeEach(() => {
      $tree = $('#opened');
    });

    it('specifies opened nodes', () => {
      expect($tree.find('li[data-node-id="1"]').is(':visible')).toEqual(true);
      expect($tree.find('li[data-node-id="1.1"]').is(':visible')).toEqual(true);
      expect($tree.find('li[data-node-id="1.1.1"]').is(':visible')).toEqual(true);
      expect($tree.find('li[data-node-id="1.2.1"]').is(':visible')).toEqual(false);
    });
  });

  describe('callbacks', () => {
    let $tree;
    let $message;

    beforeEach(() => {
      $tree = $('#callback');
      $message = $('#message');
    });
    
    it('runs callbacks', () => {
      $tree.find(`li[data-node-id="1.1"] .${NAMESPACE}-icon`).click().click();
      expect($message.html()).toContain("opened 1.1");
      expect($message.html()).toContain("closed 1.1");
    });
  });

  describe('destroy', () => {
    let $tree;

    beforeEach(() => {
      eval($('script').text());
      $tree = $('#callback');
    });
    
    it('destroy', () => {
      $tree.data(NAMESPACE).destroy();
      expect($tree.hasClass(NAMESPACE)).toEqual(false);
      expect($._data($tree.get(0), 'events')).toEqual(undefined);
    });
  });
});


test/jquery-simple-tree-config-spec.js
describe('jquery-simple-tree', () => {
  it('config', () => {
    let defaults = $.SimpleTree.getDefaults();
    expect(defaults.expander).toEqual(null);

    defaults = $.SimpleTree.setDefaults({expander: 'test'});
    expect(defaults.expander).toEqual('test');
  });
});


src/store.js
import $ from 'jquery';

export default class Store {
  constructor(tree, options = {}) {
    this.tree = tree;
    this.key = options.storeKey;
    this.storage = Store.storage(options.storeType);
  }

  save() {
    let ids = this.tree.nodes().filter('.tree-opened').map((i, node) => {
      return $(node).data('node-id');
    }).get();

    Store.saveData(this.storage, this.key, ids);
  }

  load() {
    let ids = Store.loadData(this.storage, this.key);
    if (!ids) {
      return;
    }

    this.tree.nodes().each((i, node) => {
      let $node = $(node);
      if (ids.indexOf($node.data('node-id')) != -1) {
        this.tree.show($node);
      } else {
        this.tree.hide($node);
      }
    });
  }

  static storage(type) {
    if (type == 'local') {
      return window.localStorage;
    } else {
      return window.sessionStorage;
    }
  }

  static saveData(storage, key, data) {
    let json = JSON.stringify(data);
    storage.setItem(key, json);
  }

  static loadData(storage, key) {
    let json = storage.getItem(key);
    if (!json) {
      return null;
    } else {
      return JSON.parse(json);
    }
  }
}


karma.conf.js
// Karma configuration
// Generated on Fri Nov 23 2018 10:20:11 GMT+0900 (JST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'node_modules/jquery/dist/jquery.min.js',
      'src/*.js',
      'test/*spec.js',
      'index.html', {
        pattern: 'dist/*',
        served: true,
        included: false
      }
    ],

    proxies: {
      '/dist/': '/base/dist/',
    },


    // list of files / patterns to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'src/*.js': ['webpack', 'sourcemap'],
      'test/*spec.js': ['webpack', 'sourcemap'],
      'index.html': ['html2js']
    },

    webpack: (function() {
      var webpack = require('./webpack.config');
      webpack.mode = 'development';
      webpack.devtool = 'inline-source-map';
      webpack.module.rules.push({
        test: /\.js$/,
        include: /src/,
        exclude: /node_modules/,
        use: {
          loader: 'coverage-istanbul-loader'
        }
      });
      return webpack;
    })(),


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress', 'coverage-istanbul'],

    coverageIstanbulReporter: {
      reports: ['html', 'lcovonly'],
      dir: __dirname + '/coverage'
    },

    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: (function() {
      if (process.platform === "win32") {
        return ['Chrome', 'Firefox', 'IE', 'Edge']
      } else {
        return ['ChromeHeadless', 'FirefoxHeadless']
      }
    })(),


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    // concurrency: Infinity
    concurrency: 1
  })
}


webpack.config.js
var webpack = require("webpack");

module.exports = {
  mode: "production",

  entry: {
    "jquery-simple-tree": "./src/jquery-simple-tree.js"
  },

  output: {
    path: __dirname + "/dist",
    filename: "[name].js",
    publicPath: "/dist"
  },

  externals: {
    jquery: "jQuery"
  },

  module: {
    rules: [
      {
        test: /\.js$/,
        include: [
          /src/,
          /test/,
          /node_modules\/@kanety\/js-store/
        ],
        use: {
          loader: "babel-loader",
          options: {
            presets: ["@babel/preset-env"]
          }
        }
      }, {
        test: /\.scss$/,
        use: [
          "style-loader",
          "css-loader",
          "sass-loader"
        ]
      }
    ]
  },

  watchOptions: {
    poll: 1000
  },

  devServer: {
    host: "0.0.0.0",
    port: 3000,
    disableHostCheck: true
  }
};


src/simple-tree.js
import $ from 'jquery';
import Store from '@kanety/js-store';

import './simple-tree.scss';
import { NAMESPACE } from './consts';

const DEFAULTS = {
  expander: null,
  collapser: null,
  opened: 'all',
  iconTemplate: '<a href="#" />',
  store: null,
  storeKey: null
};

export default class SimpleTree {
  constructor(element, options = {}) {
    this.options = $.extend({}, DEFAULTS, options);

    this.$root = $(element);
    this.$expander = $(this.options.expander);
    this.$collapser = $(this.options.collapser);

    if (this.options.store && this.options.storeKey) {
      this.store = new Store({
        type: this.options.store,
        key: this.options.storeKey
      })
    }

    this.init();
    this.load();
  }

  init() {
    this.$root.addClass(NAMESPACE);
    this.build();
    this.unbind();
    this.bind();
  }

  destroy() {
    this.$root.removeClass(NAMESPACE);
    this.$root.off(`node:open node:close`);
    this.unbuild();
    this.unbind();
  }

  build() {
    this.nodes().each((i, node) => {
      let $node = $(node);

      if ($node.children(`.${NAMESPACE}-icon`).length == 0) {
        let $icon = $(this.options.iconTemplate).addClass(`${NAMESPACE}-icon`);
        $node.prepend($icon);
      }

      let cls = $node.attr('class');
      let reg = new RegExp(`^${NAMESPACE}-(opened|closed)$`);
      if (!cls || !cls.split(' ').some(str => str.match(reg))) {
        if ($node.children('ul').length == 0) {
          if ($node.data('node-has-children')) {
            $node.addClass(`${NAMESPACE}-closed`);
          } else {
            $node.addClass(`${NAMESPACE}-closed ${NAMESPACE}-hidden-icon`);
          }
        } else if ($node.data('node-keep-open')) {
          $node.addClass(`${NAMESPACE}-opened ${NAMESPACE}-hidden-icon`);
          $node.children('ul').show();
        } else if (this.opensDefault($node)) {
          $node.addClass(`${NAMESPACE}-opened`);
          $node.children('ul').show();
        } else {
          $node.addClass(`${NAMESPACE}-closed`);
          $node.children('ul').hide();
        }
      }
    });
  }

  unbuild() {
    let detector = (i, className) => {
      let reg = new RegExp(`${NAMESPACE}(-\\S+)?`, 'g');
      return (className.match(reg) || []).join(' ');
    }
    this.nodes().removeClass(detector);
    this.$root.find(`.${NAMESPACE}-icon`).remove();
  }

  opensDefault($node) {
    let opened = this.options.opened;
    return opened && (opened == 'all' || opened.indexOf($node.data('node-id')) != -1);
  }

  bind() {
    this.$expander.on(`click.${NAMESPACE}`, (e) => {
      e.preventDefault();
      this.expand()
    });
    this.$collapser.on(`click.${NAMESPACE}`, (e) => {
      e.preventDefault();
      this.collapse()
    });

    this.$root.on(`click.${NAMESPACE}`, `.${NAMESPACE}-icon`, (e) => {
      e.preventDefault();
      let $node = $(e.currentTarget).parent();
      if (this.isOpened($node)) {
        this.close($node);
      } else {
        this.open($node);
      }
    }).on(`keydown.${NAMESPACE}`, `.${NAMESPACE}-icon`, (e) => {
      let $node = $(e.currentTarget).parent();
      if (this.keydown($node, e.keyCode)) {
        e.preventDefault();
      }
    });
  }

  unbind() {
    this.$expander.off(`.${NAMESPACE}`);
    this.$collapser.off(`.${NAMESPACE}`);
    this.$root.off(`.${NAMESPACE}`);
  }

  expand() {
    this.nodes().each((i, node) => {
      this.show($(node));
    });
    this.save();
  }

  collapse() {
    this.nodes().each((i, node) => {
      this.hide($(node));
    });
    this.save();
  }

  nodes() {
    return this.$root.find('li');
  }

  icons() {
    return this.$root.find(`.${NAMESPACE}-icon`);
  }

  visibleIcons() {
    let $icons = this.icons().filter(':visible');
    return $icons.filter((index) => $icons.eq(index).css('visibility') != 'hidden');
  }

  iconOf($node) {
    return $node.children(`.${NAMESPACE}-icon`).first();
  }

  open($node) {
    this.show($node);
    this.save();

    $node.trigger('node:open', [$node]);
  }

  close($node) {
    this.hide($node);
    this.save();

    $node.trigger('node:close', [$node]);
  }

  show($node) {
    if ($node.hasClass(`${NAMESPACE}-hidden-icon`)) return;
    if ($node.data('node-keep-open')) return;

    $node.removeClass(`${NAMESPACE}-closed`).addClass(`${NAMESPACE}-opened`);
    $node.children('ul').show();
  }

  hide($node) {
    if ($node.hasClass(`${NAMESPACE}-hidden-icon`)) return;
    if ($node.data('node-keep-open')) return;

    $node.removeClass(`${NAMESPACE}-opened`).addClass(`${NAMESPACE}-closed`);
    $node.children('ul').hide();
  }

  isOpened($node) {
    return $node.hasClass(`${NAMESPACE}-opened`);
  }

  isClosed($node) {
    return $node.hasClass(`${NAMESPACE}-closed`);
  }

  hasChildren($node) {
    return $node.children('ul').length > 0;
  }

  keydown($node, keyCode) {
    switch (keyCode) {
    case 37: // left
      this.keyLeft($node);
      return true;
    case 38: // up
      this.keyUp($node);
      return true;
    case 39: // right
      this.keyRight($node);
      return true;
    case 40: // down
      this.keyDown($node);
      return true;
    }
    return false;
  }

  keyUp($node) {
    let $icons = this.visibleIcons();
    let index = $icons.index(this.iconOf($node)) - 1;
    if (index >= 0) $icons.eq(index).focus();
  }

  keyDown($node) {
    let $icons = this.visibleIcons();
    let index = $icons.index(this.iconOf($node)) + 1;
    if (index >= 0) $icons.eq(index).focus();
  }

  keyRight($node) {
    if (this.isOpened($node)) {
      this.iconOf($node.find(`> ul > li`)).focus();
    } else {
      this.open($node);
    }
  }

  keyLeft($node) {
    if (this.isClosed($node)) {
      this.iconOf($node.parent().parent()).focus();
    } else {
      this.close($node);
    }
  }

  findByID(id) {
    return this.$root.find(`li[data-node-id="${id}"]`).first();
  }

  openByID(id) {
    this.open(this.findByID(id));
  }

  closeByID(id) {
    this.close(this.findByID(id));
  }

  save() {
    if (!this.store) return;

    let ids = this.nodes().filter(`.${NAMESPACE}-opened`).map((i, node) => {
      return $(node).data('node-id');
    }).get();

    this.store.set(ids);
  }

  load() {
    if (!this.store) return;

    let ids = this.store.get();
    if (!ids) return;

    this.nodes().each((i, node) => {
      let $node = $(node);
      if (ids.indexOf($node.data('node-id')) != -1) {
        this.show($node);
      } else {
        this.hide($node);
      }
    });
  }

  static getDefaults() {
    return DEFAULTS;
  }

  static setDefaults(options) {
    return $.extend(DEFAULTS, options);
  }
}
